Settings can understand and parse trustedSigners

PatoBeltran opened the pull request:
## Bug
First part of https://github.com/NuGet/Home/issues/6961 

Adds the ability to understand, validate and parse correctly `trustedSigners` and `signatureVerificationMode` from nuget.config.

PatoBeltran committed:
settings can understand and parse trustedSigners

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Resources.resx:
```diff
@@ -221,6 +227,12 @@
+    <value>A trusted signer entry must have at least one certificate entry.</value>
+  </data>
+  <data name="UnknownHashAlgorithmNotSupported" xml:space="preserve">
+    <value>Unknown hash algorithm is not supported.</value>
```
Would it be helpful if we call out which value that is?

PatoBeltran replied:
This is only thrown when the item is created in memory (basically the user passed `HashAlgorithm.Unknown` to the method), when parsed we do say which value is the invalid one/

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Resources.resx:
```diff
@@ -196,6 +199,9 @@
     <value>Environment variable '{0}' must contain an absolute path, the full path of '{1}' cannot be determined.</value>
   </data>
+  <data name="RepositoryMustHaveOneOwners" xml:space="preserve">
+    <value>A repository item can only have one owners item in it.</value>
```
Can you say which item?

PatoBeltran replied:
Good idea!

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/AuthorItem.cs:
```diff
@@ -0,0 +1,103 @@
+
+        internal override SettingBase Clone()
+        {
+            var newItem = new AuthorItem(Name, Certificates.ToArray());
```
What's the intention of the cloning here? 
This is not a deep clone as Certificates itself is mutable.

PatoBeltran replied:
You are right, certificates should also be cloned. This is to follow the pattern of all `SettingItem`. Cloning is used when getting the values to make sure the user gets copies instead of the values that hold the reference to the XElements

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/AuthorItem.cs:
```diff
@@ -0,0 +1,103 @@
+            return string.Equals(Name, other.Name, StringComparison.Ordinal);
+        }
+
+        public bool DeepEquals(AuthorItem other)
```
Should these be internal/private if you're just gonna invoke them through the overrides?

PatoBeltran replied:
it could be made private, but it is this way for all other `SettingItems` do you think we should update this in all of them?

nkolev92 replied:
I'd say unless you have a better reason not to. 

It doesn't have to be this PR. 
It can be a side PR that we can quickly go through.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/CertificateItem.cs:
```diff
@@ -0,0 +1,115 @@
+            return string.Equals(Fingerprint, other.Fingerprint, StringComparison.Ordinal);
+        }
+
+        public bool DeepEquals(CertificateItem other)
```
Similar to the earlier item class. 
Should these be private?

nkolev92 reviewed (approved):
LGTM. 
Some small comments.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/TrustedSignerItem.cs:
```diff
@@ -0,0 +1,79 @@
+
+        private IList<CertificateItem> _certificates { get; }
+
+        public IReadOnlyList<CertificateItem> Certificates => _certificates.ToList().AsReadOnly();
```
`_certificates` is already a list so you dont need to do `ToList()`

PatoBeltran replied:
`IList` does not implement `IReadOnlyList` nor has an implementation for `AsReadOnly()` theferore both `ToList()` and `.AsReadOnly()` are needed.

dtivel replied:
Redefine `_certificates` as:

    private readonly List<CertificateItem> _certificates;

`List<T>` does implement `IReadOnlyList<T>`, and it is just fine to retain the concrete type for a completely private field.

Then simplify this property to:

    public IReadOnlyList<CertificateItem> Certificates => _certificates;

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/TrustedSignerItem.cs:
```diff
@@ -0,0 +1,79 @@
+
+            AddAttribute(ConfigurationConstants.NameAttribute, name);
+
+            _certificates = new List<CertificateItem>();
```
why create a new list? Just check for null or empty then assign as a list to _certificates

PatoBeltran replied:
To avoid null checking everywhere else I want to make sure there is always a list (even if it's empty). This is the same pattern followed in `UnknownItem`, `SettingGroup` and `SettingElement`.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/TrustedSignerItem.cs:
```diff
@@ -0,0 +1,79 @@
+
+namespace NuGet.Configuration
+{
+    public abstract class TrustedSignerItem : SettingItem
```
I dont understand why some items extend from `IEquatable` vs some not... Should this be consistent across your `SettingItem`?

PatoBeltran replied:
All sealed classes that inherit from `SettingItem` implement `IEquatable`

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/CertificateItem.cs:
```diff
@@ -0,0 +1,115 @@
+
+            AddAttribute(ConfigurationConstants.Fingerprint, fingerprint);
+            AddAttribute(ConfigurationConstants.HashAlgorithm, hashAlgorithm.ToString());
+            AddAttribute(ConfigurationConstants.AllowUntrustedRoot, $"{allowUntrustedRoot}");
```
shouldn't allowUntrustedRoot.toString() work for you?

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/CertificateItem.cs:
```diff
@@ -0,0 +1,115 @@
+            }
+        }
+
+        protected override HashSet<string> RequiredAttributes => new HashSet<string>() { ConfigurationConstants.Fingerprint, ConfigurationConstants.HashAlgorithm, ConfigurationConstants.AllowUntrustedRoot };
```
what is the purpose of this `RequiredAttributes`?

PatoBeltran replied:
When parsing an item, if it doesn't have any of the required attributes, parsing will throw because the item is not valid.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/CertificateItem.cs:
```diff
@@ -0,0 +1,115 @@
+        {
+            get
+            {
+                if (bool.TryParse(Attributes[ConfigurationConstants.AllowUntrustedRoot], out var parsedAttribute))
```
isn't `AllowUntrustedRoot` always be there for Certificate item?

PatoBeltran replied:
yes, but it might be the case that a user hand edited the file and added something different than `true` or `false`, therefore I want to make sure I parse it correctly.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/CertificateItem.cs:
```diff
@@ -0,0 +1,115 @@
+            return newItem;
+        }
+
+        public bool Equals(CertificateItem other)
```
May be I'm missing something... but why there are different impl for Equals and DeepEquals? I didnt even know there is some terminology as DeepEquals :)

PatoBeltran replied:
For any `SettingItem`: two items that have `Equals` true means that both of those items refer to the same setting, this does not mean that the items have the exact same values (this is what `DeepEquals` is for).

Think for example of an `AddItem`... `<add key='key1' value='val' /> and `<add key='key1' value='differentVal' /> would have `Equals() == true` but `DeepEquals() == false`.

jainaashish replied:
traditionally `Equals` is considered as "deep equals" so it's little confusing. So may be change `Equals` to `EqualsKey` or `KeyEquals` or something like that...

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/AuthorItem.cs:
```diff
@@ -0,0 +1,103 @@
+
+namespace NuGet.Configuration
+{
+    public sealed class AuthorItem : TrustedSignerItem, IEquatable<AuthorItem>
```
why do you need separate class impl for author vs repo signers? Can it be part of `TrustedSignerItem` itself with a type and ServiceIndex being optional for author...

PatoBeltran replied:
I think it is cleaner to make each class have a single responsibility. By making the implementation only one class and validating you are doing some things for authors and other for repositories the code becomes more error-prone. It is cleaner and easier to maintain if we separate that responsibility in two different classes that subclass from the same one.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/AuthorItem.cs:
```diff
@@ -0,0 +1,103 @@
+        {
+        }
+
+        internal override SettingBase Clone()
```
Return type should be `AuthorItem`

PatoBeltran replied:
This is an override from `SettingBase` the type cannot be overwritten.

jainaashish replied:
this doesn't look good then may be dont override clone(0..?

nkolev92 replied:
You can have a generic class that SettingsBase and the rest implement. 
That way the return type of clone can be the type of the class itself. 

We already have ICloneable too, but I don't like ICloneable cause it doesn't enforce any contracts.

jainaashish reviewed (commented):
Still going through the PR...

dtivel commented on src/NuGet.Core/NuGet.Common/SignatureValidationMode.cs:
```diff
@@ -0,0 +1,21 @@
+    public enum SignatureValidationMode
+    {
+        /// <summary>
+        /// Relaxed siganture verification mode.
```
signature

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/TrustedSignerItem.cs:
```diff
@@ -0,0 +1,79 @@
+
+        protected override bool CanHaveChildren => true;
+
+        private IList<CertificateItem> _certificates { get; }
```
Use a private readonly field instead.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/AuthorItem.cs:
```diff
@@ -0,0 +1,103 @@
+    {
+        public override string ElementName => ConfigurationConstants.Author;
+
+        protected override HashSet<string> RequiredAttributes => new HashSet<string>() { ConfigurationConstants.NameAttribute };
```
I don't see this used anywhere.

This property will always return a new `HashSet<string>()` instance with one element, which I think is wrong.

I think you should return the same instance, preferrably as an immutable collection:

    protected override IReadOnlyCollection<string> RequiredAttributes { get; } = new HashSet<string>() { ConfigurationConstants.NameAttribute };

If it must be mutable, at least change `=>` to `{ get; } =`.

PatoBeltran replied:
This is used [here](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Configuration/Settings/SettingElement.cs#L252). `AddItem` also overrides this (and potentially any item can) . I can create another PR to change this in `AddItem` and make it an `IReadOnlyCollection<string>`  in `SettingElement`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/TrustedSignerItem.cs:
```diff
@@ -0,0 +1,79 @@
+
+            _certificates = new List<CertificateItem>();
+
+            if (certificates == null || !certificates.Any())
```
Move this validation immediately after the `name` validation.  Perform argument validation upfront.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/TrustedSignerItem.cs:
```diff
@@ -0,0 +1,79 @@
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentNullException(nameof(name));
```
[This old feedback](https://github.com/NuGet/NuGet.Client/pull/2370#discussion_r217167772) applies to this PR as well.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/AuthorItem.cs:
```diff
@@ -0,0 +1,103 @@
+
+        internal override XNode AsXNode()
+        {
+            if (Node != null && Node is XElement)
```
You don't need the explicit `null` check.  `is` includes a `null` check.  See https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/is.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/CertificateItem.cs:
```diff
@@ -0,0 +1,115 @@
+    {
+        public string Fingerprint => Attributes[ConfigurationConstants.Fingerprint];
+
+        public HashAlgorithmName HashAlgorithm => CryptoHashUtility.GetHashAlgorithmName(Attributes[ConfigurationConstants.HashAlgorithm]);
```
`CryptoHashUtility.GetHashAlgorithmName(...)` is going to be called every time this property is accessed.

Why not just set this in the constructor?

    public HashAlgorithmName HashAlgorithm { get; }

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/CertificateItem.cs:
```diff
@@ -0,0 +1,115 @@
+            }
+        }
+
+        protected override HashSet<string> RequiredAttributes => new HashSet<string>() { ConfigurationConstants.Fingerprint, ConfigurationConstants.HashAlgorithm, ConfigurationConstants.AllowUntrustedRoot };
```
My earlier comment on `AuthorItem.RequiredAttributes` applies here as well.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/CertificateItem.cs:
```diff
@@ -0,0 +1,115 @@
+        {
+            if (string.IsNullOrEmpty(fingerprint))
+            {
+                throw new ArgumentNullException(nameof(fingerprint));
```
My earlier comment on throwing `ArgumentException` applies here as well.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/CertificateItem.cs:
```diff
@@ -0,0 +1,115 @@
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
```
Can `DeepEquals(...)` call `Equals(...)` first?

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/CertificateItem.cs:
```diff
@@ -0,0 +1,115 @@
+
+            if (other.Attributes.Count == Attributes.Count)
+            {
+                return Attributes.OrderedEquals(other.Attributes, data => data.Key, StringComparer.OrdinalIgnoreCase);
```
Wouldn't this compare the fingerprint case-insensitively?  How does that jive with [this](https://github.com/NuGet/NuGet.Client/pull/2426/files#diff-cfe7b60de891d179cb60eb67bc62589cR87)?

PatoBeltran replied:
No, this is comparing the key case-insensitively, therefore the word "fingerprint"

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/AuthorItem.cs:
```diff
@@ -0,0 +1,103 @@
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
```
Can `DeepEquals(...)` call `Equals(...)` first?

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/OwnersItem.cs:
```diff
@@ -0,0 +1,100 @@
+
+        protected override bool CanHaveChildren => true;
+
+        public IList<string> Content => _content.Value.Split(OwnersListSeparator).Select(o => o.Trim()).ToList();
```
`IReadOnlyList<T>`?

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/OwnersItem.cs:
```diff
@@ -0,0 +1,100 @@
+
+        internal override XNode AsXNode()
+        {
+            if (Node != null && Node is XElement)
```
Don't need the explicit `null` check.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Utility/SettingsUtility.cs:
```diff
@@ -79,6 +79,18 @@ public static int GetMaxHttpRequest(ISettings settings)
             return 0;
         }
 
+        public static SignatureValidationMode GetSignatureValidationMode(ISettings settings)
```
Do you verify that `settings` is not `null`?

dtivel commented on test/NuGet.Core.Tests/NuGet.Configuration.Test/SettingsFileParsingTests/CertificateItemTests.cs:
```diff
@@ -0,0 +1,164 @@
+                var item = section.Items.First() as CertificateItem;
+
+                var expectedItem = new CertificateItem("abcdefg", Common.HashAlgorithmName.SHA256, allowUntrustedRoot: true);
+                item.DeepEquals(expectedItem).Should().BeTrue();
```
The test relies on correctness of `DeepEquals(...)` but there are no tests for `DeepEquals(...)`.  Similarly there are no tests for other members like `Equals(...)`, `ElementName`, and `Clone()`.

This feedback applies to all new settings types you're adding, not just `CertificateItem`.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Items/CertificateItem.cs:
```diff
@@ -0,0 +1,115 @@
+        internal CertificateItem(XElement element, SettingsFile origin)
+            : base(element, origin)
+        {
+            if (HashAlgorithm == HashAlgorithmName.Unknown)
```
`HashAlgorithm` hasn't been set as part of this PR. isn't it?

PatoBeltran replied:
since it is an attribute it is parsed as part of `SettingElement` creation. Here we just need to validate it's value is supported.

jainaashish reviewed (approved):
Few minor things, which either be changed or discuss before proceeding with current impl as it changes the existing meaning to these predefined APIs:
- clone() api should return the same class instance
- Equals() should be deep equals, if you need a separate equals for key then do something `KeyEquals`

PatoBeltran committed:
address feedback

PatoBeltran commented:
@nkolev92 @jainaashish @dtivel I addressed all your comments and updated the PR with more tests and the update functionality which was missing from the original PR. Please review 😄
